jobs/MeterInstanceJob.js

```diff
@@ -0,0 +1,145 @@
+      return this.runFailed(err, {}, job, done)
+    }
+  }
+  /* jshint ignore:end */
+
+  static getInstanceEvents() {
+    const options = {
+      resourceGroup: CONST.APISERVER.RESOURCE_GROUPS.INSTANCE,
+      resourceType: CONST.APISERVER.RESOURCE_TYPES.SFEVENT,
+      query: {
+        labelSelector: `state in (${CONST.METER_STATE.TO_BE_METERED},${CONST.METER_STATE.FAILED})`
+      }
+    };
+    return apiServerClient.getResources(options);
+  }
+
+  /* jshint ignore:start */
+  static async meter(events) {
+    try {
+      logger.info(`Number of events to be metered in this run - ${events.length}`);
+      // Adding this comment as we are transitioning to async/await
+      // Note: The below Promise is not bluebird promise
+      const resultArray = await Promise.all(_.map(events, async (event) =>
+        await this.sendEvent(event)));
+      const successCount = resultArray.filter(r => r === true).length;
+      return {
+        totalEvents: events.length,
+        success: successCount,
+        failed: events.length - successCount,
+      }
+    } catch (err) {
+      logger.error(`Error occurred while metering all events : `, err);
+    }
+  }
+  /* jshint ignore:end */
+
+  static isServicePlanExcluded(options) {
+    const serviceId = _.get(options, 'service.service_guid');
+    const planId = _.get(options, 'service.plan_guid');
+    logger.info(`Checking if service ${serviceId}, plan: ${planId} is excluded`);
+    const plan = catalog.getPlan(planId);
+    return !plan.metered;
+  }
+
+  static enrichEvent(options) {
+    // Add region , service name and plan sku name of the event
+    const serviceId = _.get(options, 'service.service_guid');
+    const planId = _.get(options, 'service.plan_guid');
+    logger.info(`Enriching the metering event ${serviceId}, plan: ${planId}`);
+    options.service.id = catalog.getServiceName(serviceId);
+    options.service.plan = catalog.getPlanSKUFromPlanGUID(serviceId, planId);
+    options.service = _.omit(options.service, ['service_guid', 'plan_guid']);
+    options.consumer.region = config.metering.region;
+    return options;
+  }
+
+  /* jshint ignore:start */
+  static async sendEvent(event) {
+    try {
+      logger.debug('Metering Event details before enriching:', event);
+      if (this.isServicePlanExcluded(event.spec.options) === true) {
+        await this.updateMeterState(CONST.METER_STATE.EXCLUDED, event);
+        return true;
+      }
+      const enrichedUsageDoc = await this.enrichEvent(_.get(event.spec, 'options'));
+      logger.info('Sending enriched document:', enrichedUsageDoc);
+      const validEvent = await maas.client.sendUsageRecord({
+        usage: [enrichedUsageDoc]
+      });
+      if (validEvent !== undefined) {
+        this.updateMeterState(CONST.METER_STATE.METERED, event);
+      }
+      return true;
+    } catch (err) {
+      logger.error('Error occured while metering:', err);
+      await MeterInstanceJob._logMeteringEvent(err, event);
```

Comment:
Are we logging every meter instance failure attempt to Riemann? What will be the strategy to raise AVS alerts? Do we flag each failure to update as downtime?

Reply:
`_logMeteringEvent` has internal logic to raise events only on failures that occur after `thresholdHours`. Hence we can raise all Riemann logs as alerts.
As long as the instance is usable we need not flag it as downtime.